### PR TITLE
feat(frontend): binder field selection

### PR DESCRIPTION
**Pull request by cykbls01**
## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***
1. Use column_desc to replace column_name and data_type in column_binding, this goal is to get the field_descs when binding the column.
2. Implement binder field selection in below format.
``` 
select (table.v1).v2 from table;
select (v1).v2 from table;
select (table.v1).* from table;
select (v1).* from table;
select (v1).v2 from (select v1 from table);
``` 

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
https://github.com/singularity-data/risingwave/issues/580
https://github.com/singularity-data/risingwave/issues/583

### CODE REVIEW

**Comment by st1page** [comment]
a little confusing for there are too many `ColumnDesc` without a valid ColmnId in our binder and other places.

**Comment by neverchanje** [comment]
I think so, it will mislead in case people use the column id and don't know it's invalid.

File: src/frontend/src/binder/window_table_function.rs [hunk 1]
```diff
@@ -109,22 +111,27 @@ impl Binder {
...
-                    ("window_start".to_string(), DataType::Timestamp, false),
-                    ("window_end".to_string(), DataType::Timestamp, false),
-                ]
-                .into_iter(),
-            );
+        let columns = columns.iter().cloned().chain(
+            [
+                ColumnCatalog {
+                    column_desc: ColumnDesc::with_named(
+                        ColumnId::new(0),
```
[the remark this answers is not in the document]
Comment (cykbls01) [reply src/frontend/src/binder/window_table_function.rs]: No, this is just a default value and will not use after in this situation.
Maybe add a function with default column_id is better.

File: src/frontend/src/binder/bind_context.rs [hunk 2]
```diff
@@ -84,6 +76,18 @@ impl BindContext {
...
 
+    pub fn get_column_binding(
+        &self,
+        table_name: Option<String>,
+        column_name: &String,
+    ) -> Result<&ColumnBinding> {
+        let index = match table_name {
+            Some(table_name) => self.get_index_with_table_name(column_name, &table_name)?,
+            None => self.get_index(column_name)?,
+        };
```
[the remark this answers is not in the document]
Comment (cykbls01) [reply src/frontend/src/binder/bind_context.rs:87]: The tabe name in get_column_binding_index is Option<&String>,
but the table name in get_column_binding is Option<String>.
So I first match the tabe_name.

**Comment by neverchanje** [comment]
Well pls move the struct-related binding to a separate file `struct_field.rs` thus we will be able to know which function is privately used, e.g `bind_field` apparently should not be public.

File: src/frontend/src/binder/select.rs [hunk 3]
```diff
@@ -153,35 +192,185 @@ impl Binder {
...
+                        )
+                        .into()
+                    })
+                    .collect_vec(),
+                desc,
+            ))
+        }
+    }
+
+    pub fn bind_columns_iter<'a>(
```
Comment (neverchanje) [inline src/frontend/src/binder/select.rs:364]:
```suggestion
    pub fn iter_bound_columns<'a>(
```

File: src/frontend/src/binder/select.rs [hunk 4]
```diff
@@ -153,35 +192,185 @@ impl Binder {
...
+        }
+    }
+
+    /// Will fail if it's an atomic value.
+    /// Rewrite (expr:Struct).* to [Field(expr, 0), Field(expr, 1), ... Field(expr, n)].
+    pub fn bind_wildcard_field(
+        expr: ExprImpl,
+        desc: ColumnDesc,
+    ) -> Result<(Vec<ExprImpl>, ColumnDesc)> {
+        if desc.field_descs.is_empty() {
```
Comment (neverchanje) [inline src/frontend/src/binder/select.rs:336]: This is not always correct. Assume an empty schema:

```
message Foo {
}
```

Will postgres fail on `foo.*`?

**Review by neverchanje — APPROVED** [review]

**Pull request merged**